landlab/components/snow/snow_degree_day.py

```diff
@@ -0,0 +1,361 @@
+    --------
+    >>> import numpy as np
+    >>> from landlab import RasterModelGrid
+    >>> from landlab.components.snow import SnowDegreeDay
+
+    >>> grid = RasterModelGrid((2, 2))
+
+    >>> grid.add_zeros("atmosphere_water__precipitation_leq-volume_flux", at="node")
+    array([0., 0., 0., 0.])
+    >>> grid.at_node["atmosphere_bottom_air__temperature"] = [2.0, 2.0, 1.0, -1.0]
+    >>> grid.at_node["snowpack__liquid-equivalent_depth"] = [0.005, 0.01, 0.005, 0.005]
+
+    >>> sm = SnowDegreeDay(grid, c0=3.0, threshold_temp=0.0)
+    >>> sm.run_one_step(8.64e4)  # 1 day in sec
+
+    >>> grid.at_node["snowpack__melt_volume_flux"] * 86400.0
+    array([0.006, 0.006, 0.003, 0.   ])
+    >>> grid.at_node["snowpack__liquid-equivalent_depth"]
+    array([0.   , 0.004, 0.002, 0.005])
+
+    >>> sm.total_snow_precip_at_node
+    array([0., 0., 0., 0.])
+    >>> sm.total_snow_melt_at_node
+    array([0.005, 0.006, 0.003, 0.   ])
+    """
+
+    _name = "SnowDegreeDay"
+
+    _unit_agnostic = False
+
+    _info = {
+        # input fields
+        "atmosphere_bottom_air__temperature": {
+            "dtype": float,
+            "intent": "in",
+            "optional": False,
+            "units": "deg_C",
+            "mapping": "node",
+            "doc": "atmosphere bottom air temperature",
+        },
+        "atmosphere_water__precipitation_leq-volume_flux": {
+            "dtype": float,
+            "intent": "in",
+            "optional": False,
+            "units": "m/s",
+            "mapping": "node",
+            "doc": "precipitation rate (in water equivalent)",
+        },
+        "snowpack__liquid-equivalent_depth": {
+            "dtype": float,
+            "intent": "inout",
+            "optional": True,
+            "units": "m",
+            "mapping": "node",
+            "doc": "snow water equivalent depth",
+        },
+        "snowpack__z_mean_of_mass-per-volume_density": {
+            "dtype": float,
+            "intent": "in",
+            "optional": True,
+            "units": "kg m-3",
+            "mapping": "node",
+            "doc": "snow density",
+        },
+        # output fields
+        "snowpack__depth": {
+            "dtype": float,
+            "intent": "out",
+            "optional": False,
+            "units": "m",
+            "mapping": "node",
+            "doc": "snow depth",
+        },
+        "snowpack__melt_volume_flux": {
+            "dtype": float,
+            "intent": "out",
+            "optional": False,
+            "units": "m/s",
+            "mapping": "node",
+            "doc": "snow melt volume flux",
+        },
+    }
+
+    def __init__(
+        self, grid, c0=2, threshold_temp=0.0, rho_water=1000.0, rain_snow_temp=1.0
+    ):
+        """Initialize SnowDegreeDay component."""
+
+        super().__init__(grid)
+
+        self.c0 = c0
+        self.threshold_temp = threshold_temp
+        self.rho_water = rho_water
+        self.rain_snow_temp = rain_snow_temp
```

Comment:
These line are using the setters, which then set the private versions of the attributes. It's, admittedly, confusing because the private versions aren't explicitly set within the class's ``__init__`` method. I wrote it this way so I wouldn't have to rewrite some of the checks that ensure the parameters are positive.